homeassistant/components/awair/config_flow.py

```diff
@@ -0,0 +1,104 @@
+
+        await self._abort_if_configured(conf[CONF_ACCESS_TOKEN])
+
+        user, errors = await self._check_access_token(conf[CONF_ACCESS_TOKEN])
+        if errors is not None:
+            return self.async_abort(reason="auth")
+
+        return self.async_create_entry(
+            title=f"{user.email} ({user.user_id})",
+            data={CONF_ACCESS_TOKEN: conf[CONF_ACCESS_TOKEN]},
+        )
+
+    async def async_step_user(self, user_input: Union[dict, None] = None):
+        """Handle a flow initialized by the user."""
+        errors = {}
+
+        if user_input is not None:
+            user, errors = await self._check_access_token(user_input[CONF_ACCESS_TOKEN])
+
+            if not errors:
+                await self._abort_if_configured(user_input[CONF_ACCESS_TOKEN])
+                title = f"{user.email} ({user.user_id})"
+                return self.async_create_entry(title=title, data=user_input)
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema({vol.Required(CONF_ACCESS_TOKEN): str}),
+            errors=errors,
+        )
+
+    async def async_step_reauth(self, user_input: Union[dict, None] = None):
+        """Handle re-auth if token invalid."""
+        errors = {}
+
+        if user_input is not None:
+            if "config_entry" in user_input:
+                if self.unique_id is None:
+                    await self.async_set_unique_id(user_input["config_entry"].unique_id)
+            elif CONF_ACCESS_TOKEN in user_input:
+                access_token = user_input[CONF_ACCESS_TOKEN]
+                user, errors = await self._check_access_token(access_token)
+                if not errors:
+                    return self.async_create_entry(
+                        title=f"{user.email} ({user.user_id})",
+                        data={CONF_ACCESS_TOKEN: user_input[CONF_ACCESS_TOKEN]},
+                    )
+
+        return self.async_show_form(
+            step_id="reauth",
+            data_schema=vol.Schema({vol.Required(CONF_ACCESS_TOKEN): str}),
+            errors=errors,
+        )
+
+    async def _check_access_token(self, access_token: str):
+        """Check the access token is valid."""
+        session = async_get_clientsession(self.hass)
+        awair = Awair(access_token=access_token, session=session)
+
+        try:
+            user = await awair.user()
+            devices = await user.devices()
+            if not devices:
+                return self.async_abort(reason="no_devices")
+
+            return (user, None)
+
+        except AuthError:
+            return (None, {CONF_ACCESS_TOKEN: "auth"})
+        except AwairError as err:
+            LOGGER.error("Unexpected API error: %s", err)
+            return (None, {"base": "unknown"})
+
+    async def _abort_if_configured(self, access_token: str):
+        """Abort if this access_token has been set up."""
+        await self.async_set_unique_id(access_token)
```

Comment:
Should we really set the token as unique_id? What if the same user changes token? Is that possible? Maybe use the user_id as unique_id instead?

Reply:
It's a good idea, and unfortunately it's one that doesn't work: you can't get the user ID if your token is invalid in the first place. :laughing: 

That said - I do think it's subpar to use the token here. It leads to shenanigans with setting unique_id, etc... I'm open to alternatives but without a valid token all account information is unavailable.

Here's a thought: we could ask for the email address on the account from the user, even though it's not needed - and use something like `f"awair-{user_provided_email}"` as the unique id for the config entry.

What do you think about something like that?

Reply:
I've decided that requiring the user to input the associated email address is a fine workaround, even if we technically don't _need_ the email address.

I've updated the code to require it and refactored the way we're playing with unique_id here.